### PR TITLE
Site Credit Block Structure

### LINKED ISSUE #37045: FSE: create Footer Credit block for self-hosted sites
At this stage we should disregard Dotcom and plan concerns and just try to make this block work in self-hosted context. Hopefully this will make it easier to seamlessly integrate it to Atomic sites later on.

@noahtallen to add more details on the feasibility of the approach or replace with other issues if needed.

Please consult the parent issue for designs. It should contain everything except the Dotcom specific upgrade nudges: https://github.com/Automattic/Dotcom-roadmap/issues/50#issuecomment-544657971.

### PR DESCRIPTION
**Pull request by noahtallen**
For now, I'm looking for feedback on if this looks like the right direction for the block. We'll continue using the site option to store the data on dotcom, but the main changes we'll need (for the future) are adding integration points to override settings in a dotcom context (i.e. hooks for whether or not the block is enabled, messages to display prompting for an upgrade, etc)

#### To do for this PR:
- [x] Allow users to align the block left/right/center
- [x] Use the footercredit option for dynamic rendering in PHP.
- [x] Update Varia to not render a footer credit if FSE is enabled. (i.e. relying on the block instead)
- [x] (possibly for followup) Refactor some of the `useSiteOptions` code so it doesn't need to have so many duplicated props.

#### Changes proposed in this Pull Request
* Create site credit block files and initial structure.
* Register `footercredit` as a setting so we can access it from the API. This is the same option name used to store the data currently on .com, so we'll be able to integrate easily with the existing solution.

Current functionality:

_Note: rendering based on the selection isn't working yet. For now, test the block editor code._

#### Testing instructions
* Pull this, insert the site credit block, and play around with it. Right now, the option is persisted correctly, but it is not rendered.
* Look at the code.
* If you insert the credit block into the footer and also use this theme PR: https://github.com/Automattic/themes/pull/1612, only the block version should show on the front end.

Resolves #37045.

### CODE REVIEW

**Comment by noahtallen** [comment]
As a bonus feature, I added alignment options so it could fit with whatever design the user may want to have. Let me know if we want to keep this

**Review by mattwiebe — COMMENTED** [review]
This looks generally good so far. I find the UX of picking a logo but only seeing it once the block loses focus a bit weird. Maybe the dropdown should be an extra line underneath, not directly in the same place as where the logo/text would go?

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.php
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.php]: Yep, you're right, I copied from Varia. :) This should be slightly updated now, but I didn't change much in this area. I need to come up with a more robust way to handle all of the possibilities.

**Comment by noahtallen** [comment]
I updated to use the normal alignment buttons (not text align, block align) so that we can use the existing alignment classes on the front end. This seems to work pretty well, but some block editor interactions for right/left alignment are a bit ugly out of the box. I imagine this will improve with time.

@mattwiebe Hm, yeah that's a good point. I followed the existing design in Dotcom-roadmap/issues/50-gh, but I'm not sure if there should be a way to preview it as you select options? cc @shaunandrews

**Comment by noahtallen** [comment]
I'm proposing this solution for extending the options from which the user can select. Cc @vindl.

1. A filter `a8c_fse_get_footer_credit_link` which can be used to change where the footer credit link takes you. Defaults to WordPress.org. (Separate from the site title, which links to the home page always.)
2. A filter `a8c_fse_update_footer_credit_options` which can be used to change which options the user selects from.

An individual option in this array has the following properties. These are all defined in PHP and then exposed through the fullSiteEditing global object in JS. This way, there is a single source of truth.
 
- `label`: The text label to show the user in the Select control.
- `value`: The "ID" of this option. This is what gets saved to the 'footercredit' option to persist the data.
- `renderType`: Used to specify if we should render an icon or text. Defaults to text.
- `renderProps`: Used to specify any props to pass into the rendering function.

There are two "consumers" of this information. First is the `edit` function for the block, and second is the `render_site_credit_block` function in PHP which renders the block dynamically on the front end.

**The Edit Function**
- The edit function can easily use the label and value in a select element the user can choose from.
- It can also easily save the `value` to the site option we're using to persist the selection.
- There is a helper function which renders text or an icon based off of renderType and renderProps.

**Dynamic PHP rendering function**
- When rendering the block, it gets the persisted option and then can easily search the array defined in PHP for the other settings associated with the value.
- It supports rendering an icon with dashicons based off of whatever is specified in `renderProps.icon`.
- It supports rendering text normally.
- It supports getting the custom credit link from the other filter above.

I think this is a pretty robust solution and should handle all of the existing footer credit options. Unfortunately, there will likely be code duplication between wpcomsh and mu plugins in order to support the WordPress.com specific options, but that's probably not avoidable unless we hardcode them into the site credit block.

**Comment by shaunandrews** [comment]
We did explore having the drop down be separate from the preview, but I’m not sure it’s a big issue. The options in the drop down should be more specific like in the design; that is the logo option should say something like “WordPress logo (image)” to help signify that it uses an image instead of text. 

Also, it looks like it’s using the WP.org logo; The WP.com logo is slightly different with only a single ring.

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (Copons) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: This is very nitpicky for such a small component, but still this would be easier to read with an early return:
```jsx
if ( ! selection ) {
  return null;
}
const { renderType, renderProps, label } = selection;
const renderer = renderType === 'icon' ? CreditIconRenderer : CreditTextRenderer;
return renderer( { label, ...renderProps } );
```

Also what about making it even more explicit?
```jsx
if ( ! selection ) {
  return null;
}
if ( 'icon' === renderType ) {
  return <Icon { ...renderProps } />;
}
return <span>{ label }</span>;
```

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js
Comment (Copons) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js]: I assume that we'll also have block alignment options (e.g. wide and full) down the line.
For the other FSE blocks I chose to use `align` for block alignment, and `textAlign` for text alignment.

Would you mind keeping it consistent here as well? 🙇

Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js]: @Copons I'll look into this. I wasn't sure which made the most sense since the reality is that we're aligning several spans in a flex container, but maybe I can get around that somehow...

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: Sure, I think making this more concise will be better :)

**Comment by noahtallen** [comment]
> with two different views handled by an isEditing state

Hm. I think I follow what you're saying but I'm not sure it really improves the experience that much. IMO that seems similarly clunky

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js]: I've changed this to `align: [ 'full', 'wide' ]` and now we can set text alignment to left/right/center. This seems to be working very well :)

I didn't keep the alignment options above because they don't seem to play well with text align _and_ it seems to mostly duplicate what text align already does. (It'd probably be weird for a user to have to set "align center" and "text align center" in order to center align it. This way, they only have to set "text align center")

**Comment by shaunandrews** [comment]
> What do y'all think of having something like the Image block, with two different views handled by an isEditing state?

We did explore a number of options, including the one you mention — and decided to move forward with the design posted above as it felt like the best (for now). There's nothing saying we can't continue to iterate later.

Here's a few explorations we did:

image

image

image

**Comment by noahtallen** [comment]
I'm now removing the in progress label, as this is mergeable and some basic extension points are available. (Note: we may want to not register this block yet until wpcom stuff is set up.)

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js
Comment (Copons) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/index.js]: Yup, this is exactly what we did for the other FSE blocks!  🙂

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (Copons) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: File names should be in `kebab-case` instead of `snakeCase`.
I can see that we have several cases of snake cases in the FSE plugin, but I guess they slipped through the reviews. 🙂

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (Copons) [inline apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: Missed this at a first read: I'm questioning the value of this export (on top of the fact that I'd avoid multiple unrelated exports in the same file).

`fullSiteEditing` is a global available everywhere, so this is basically only a shortcut for a variable destructuring.
On top of this, `defaultCreditOption` is not even used in this file.

I think we could drop this and use a simpler `fullSiteEditing.footerCreditOptions` later in this file, and the destructuring in the `edit.js`.

**Comment by Copons** [comment]
I think we should move the `useSiteOptions` refactor into its own PR, since it takes basically half of this one, and affects other FSE blocks.

**Comment by Addison-Stavlo** [comment]
I am finding that the selected option gets reset when drag-and-drop is used into or out of a parent block.  That is (as outlined in the gif below), if we select the logo option and then move the block to a different parent the option resets to the default text credit.

drag-drop-reset


On another note, I am curious if this left-most option on the toolbar is showing the correct icons.  At first it is a WordPress logo with a dropdown arrow, but on mouse-over it turns into the clockwise double arrow.  Is that intended?  I am a bit lost on what either of those logos are supposed to represent here.  

Screen Shot 2019-11-01 at 12 32 25 PM

Screen Shot 2019-11-01 at 12 32 39 PM

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: Thanks! I'll change this

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/footerCreditChoices.js]: I think that's a fair point. Previously, I was defining the option here, so this is what it evolved to after I started declaring them from PHP. I'll change it

**Comment by noahtallen** [comment]
>  the option resets to the default text credit.

@Addison-Stavlo Good catch, that's really weird! I'll take a look. 

> I am curious if this left-most option on the toolbar is showing the correct icons.

Yep, this is expected :) Gutenberg uses this to show the normal block icon, but then if you mouse over it, it shows the "transform" button. This is used to turn the block into a different block - for example, changing a heading to a paragraph block. In this case, if you chose "group" block, it should create a group block and then put the site credit block inside of it.

> I think we should move the useSiteOptions refactor into its own PR

@copons that's fair. I wanted to be able to use the new HOC here because it makes things so much easier, but probably good to review separately.

**Comment by noahtallen** [comment]
> the option resets to the default text credit.

@Addison-Stavlo really good catch here! I was able to replicate it with existing blocks as well, so this is not related to changes here. It looks like the block resets when you drag it into the group, but the option hasn't been persisted to the database yet.

**Comment by noahtallen** [comment]
@copons I have moved the site option changes to #37255 and have changed this PRs base to that branch so it can use the new function it introduces. (Should also make reviewing easier.) If that PR gets merged, I'll change the base of this branch back to master.

**Comment by noahtallen** [comment]
Tried to rebase, which turned out to have way to many conflicts, so a simple merge has gotten us back to `master` as the base branch.

**Review by Addison-Stavlo — COMMENTED** [review]
Good Stuff!

I had a few questions about the code but nothing big or show stopping except the below noted testing issue.

I found an inconsistency between testing locally and on the dotcom sandbox.   That is, if we go to update a wp_template_part (header or footer) and add the credit block, select the logo option, "update", and then "go back to page" we can see that:
- in the case of local env the logo save is respected.
- in the case of dotcom sandbox the block turns back to the default text setting.

It seems a bit different than the issue I noted earlier in the settings resetting when changing parent elements, something may not be right with the options saving on dotcom? 🤔 

Local: 
local-credit

Sandbox:
sandbox-credit

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/style.scss
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/style.scss]: I think this is a great suggestion. Ideally, we would get these styles from the theme file -- I think there, it does use rem. However, when I copied that value here, it didn't display the same size. So I think the solution is to make sure theme styles are applying correctly here so that we include the rem from that

**Comment by noahtallen** [comment]
> in the case of dotcom sandbox the block turns back to the default text setting.

Hm, that is an odd isue, I'll have to look into that.

**Review by shaunandrews — APPROVED** [review]
The Calypso.live link doesn't seem to be working for this PR (I think its a FSE thing unrelated to this specific PR), but based on the images/GIFs above this all looks fine to me.

I do question the need for the `,` (comma) between the site title and the credit — especially as we visually differentiate them already — but thats a very minor things that shouldn't block this from merging.

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/edit.js
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/edit.js]: Yeah, that's a good point. I think site credit makes more sense overall

File: apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/style.scss
Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/style.scss]: Hm probably not :D I think I was figuring out what worked here, and this ended up in the code

Comment (noahtallen) [reply apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/site-credit/style.scss]: Though I guess you could say it's more obvious what it does now :trollface:

**Review by vindl — APPROVED** [review]
As discussed on Slack, I think it's fine to merge this now provided that we remove the block registration.

Making this on Dotcom will require prerequisite Core changes to land. Landing this PR now will make follow up work easier, since we won't have to spend time keeping it up to date and rebasing.

**Pull request merged**
